typings/index.d.ts

```diff
@@ -165,12 +165,14 @@ declare namespace commander {
      *
      * @param {string} flags
      * @param {string} [description]
-     * @param {((arg1: any, arg2: any) => void) | RegExp} [fn] function or default
+     * @param {((value: string, previous: any) => void) | RegExp} [fn] function or default
      * @param {*} [defaultValue]
      * @returns {Command} for chaining
      */
-    option(flags: string, description?: string, fn?: ((arg1: any, arg2: any) => void) | RegExp, defaultValue?: any): Command;
     option(flags: string, description?: string, defaultValue?: any): Command;
+    option(flags: string, description: string, regexp: RegExp, defaultValue?: any): Command;
+    option<T>(flags: string, description: string, fn: (value: string, previous: T) => (T | void), defaultValue?: T): Command;
+    option<T>(flags: string, description: string, fn: (value: string, previous: T) => void, defaultValue: T): Command;
```

Comment:
Does the last overload ever get considered? Won't any overload that could match
`option<T>(flags: string, description: string, fn: (value: string, previous: T) => void, defaultValue: T): Command;`

already be matched by:
`option<T>(flags: string, description: string, fn: (value: string, previous: T) => (T | void), defaultValue?: T): Command;`

Reply:
If I keep the current way your collector working:
- If the collector function return void, the defaultValue MUST be provide to store parameters, because, if not this option is useless an will be keep as undefined/void all the time.
- if default value is provided, so no problem the default value can be use as a storage, and the collector function can return void.

Reply:
final version:
```javascript
   option(flags: string, description: string, regexp: RegExp, defaultValue?: any): Command;
    option<T>(flags: string, description: string, fn: (value: string, previous: T) => void, defaultValue: T): Command;
    option<T>(flags: string, description: string, fn: (value: string, previous: T) => T, defaultValue?: T): Command;
    option(flags: string, description?: string, defaultValue?: any): Command;
```

Reply:
A couple of thoughts:
1) The regular expression support is no longer documented in README and soft deprecated. 
2) I am not sure if void makes sense, after experimenting tonight.

Would just two overloads work for all expected cases? (NB: this suggestion is untested as yet.)

```
option<T>(flags: string, description: string, fn: (value: string, previous: T) => T, defaultValue?: T): Command;
option(flags: string, description?: string, defaultValue?: any): Command;
```

(I think legacy code using regular expression will not cause a TypeScript error since will match signature of second overload. Untested,)

Reply:
By experimenting I confirmed that the `defaultValue?: any` allows malformed functions if there is no default value to force the match with the first overload. Trying a tighter overload after some experimentation with TypeScript, still experimenting...

```
option<T>(flags: string, description: string, fn: (value: string, previous: T) => T, defaultValue?: T): Command;
option(flags: string, description?: string, defaultValue?: string | boolean | RegExp): Command;

```